tests/flow/test_hnswlib.py

```diff
@@ -518,11 +450,481 @@ def test_multi_range_query():
+        total_res_bf.append(res_labels_bf_range[0])
+        total_results += res_labels_bf_range[0].size
+        overall_intersection_rate += res_labels_range[0].size / res_labels_bf_range[0].size\
+            if res_labels_bf_range[0].size > 0 else 1
+    print(f"Range queries - running {num_queries} queries sequentially, average number of results is:"
+          f" {total_results/num_queries} and HNSW success rate is: {overall_intersection_rate/num_queries}."
+          f" Average query time is {total_search_time/num_queries} seconds")
+
+    # Run range queries in parallel
+    start = time.time()
+    hnsw_labels_range_parallel, _ = parallel_index.range_parallel(query_data, radius=radius)
+    total_range_query_parallel_time = time.time() - start
+    overall_intersection_rate_parallel = 0
+    for i in range(num_queries):
+        query_results_set = set(hnsw_labels_range_parallel[i])
+        query_results_set.discard(PADDING_LABEL)  # remove the irrelevant padding values
+        assert query_results_set.issubset(set(total_res_bf[i]))
+        overall_intersection_rate_parallel += len(query_results_set) / total_res_bf[i].size \
+            if total_res_bf[i].size > 0 else 1
+    print(f"Running the same {num_queries} queries in parallel, average query time is"
+          f" {total_range_query_parallel_time/num_queries}, and intersection rate is: "
+          f"{overall_intersection_rate_parallel/num_queries}")
+    assert overall_intersection_rate_parallel == overall_intersection_rate
+    print(f"Got improvement of {total_search_time/total_range_query_parallel_time} times using {n_threads} threads\n")
+    assert total_search_time/total_range_query_parallel_time >= expected_speedup
+
+
+def test_parallel_with_multi():
+    dim = 32
+    num_labels = 10000
+    per_label = 5
+    epsilon = 0.01
+    k = 10
+    num_elements = num_labels * per_label
+    metric = VecSimMetric_L2
+    data_type = VecSimType_FLOAT64
+    num_queries = 10000
+    n_threads = min(os.cpu_count(), 8)
+    expected_parallel_rate = 0.9  # we expect that at least 90% of the insert/search time will be executed in parallel
+    expected_speedup = 1 / ((1-expected_parallel_rate) + expected_parallel_rate/n_threads)  # by Amdahl's law
+
+    multi_index = create_hnsw_index(dim, num_elements, metric, data_type, m=32, ef_runtime=200,
+                                    is_multi=True)
+    parallel_multi_index = create_hnsw_index(dim, num_elements, metric, data_type, m=32, ef_runtime=200,
+                                             is_multi=True)
+    bf_params = BFParams()
+
+    bf_params.dim = dim
+    bf_params.metric = metric
+    bf_params.type = data_type
+    bf_params.initialCapacity = num_elements
+    bf_params.blockSize = num_elements
+    bf_params.multi = True
+
+    bf_index = BFIndex(bf_params)
+
+    np.random.seed(47)
+    data = np.random.random((num_labels, per_label, dim))
```

Comment:
is it OK to not specify the datatype? will its size be 8 bytes?

Reply:
yes, the default is float64